[SPARK-23375][SQL] Eliminate unneeded Sort in Optimizer

mgaido91 opened the pull request:
## What changes were proposed in this pull request?

Added a new rule to remove Sort operation when its child is already sorted.
For instance, this simple code:
```
spark.sparkContext.parallelize(Seq(("a", "b"))).toDF("a", "b").registerTempTable("table1")
val df = sql(s"""SELECT b
                | FROM (
                |     SELECT a, b
                |     FROM table1
                |     ORDER BY a
                | ) t
                | ORDER BY a""".stripMargin)
df.explain(true)
```
before the PR produces this plan:
```
== Parsed Logical Plan ==
'Sort ['a ASC NULLS FIRST], true
+- 'Project ['b]
   +- 'SubqueryAlias t
      +- 'Sort ['a ASC NULLS FIRST], true
         +- 'Project ['a, 'b]
            +- 'UnresolvedRelation `table1`

== Analyzed Logical Plan ==
b: string
Project [b#7]
+- Sort [a#6 ASC NULLS FIRST], true
   +- Project [b#7, a#6]
      +- SubqueryAlias t
         +- Sort [a#6 ASC NULLS FIRST], true
            +- Project [a#6, b#7]
               +- SubqueryAlias table1
                  +- Project [_1#3 AS a#6, _2#4 AS b#7]
                     +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(assertnotnull(input[0, scala.Tuple2, true]))._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2, true, false) AS _2#4]
                        +- ExternalRDD [obj#2]

== Optimized Logical Plan ==
Project [b#7]
+- Sort [a#6 ASC NULLS FIRST], true
   +- Project [b#7, a#6]
      +- Sort [a#6 ASC NULLS FIRST], true
         +- Project [_1#3 AS a#6, _2#4 AS b#7]
            +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true, false) AS _2#4]
               +- ExternalRDD [obj#2]

== Physical Plan ==
*(3) Project [b#7]
+- *(3) Sort [a#6 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(a#6 ASC NULLS FIRST, 200)
      +- *(2) Project [b#7, a#6]
         +- *(2) Sort [a#6 ASC NULLS FIRST], true, 0
            +- Exchange rangepartitioning(a#6 ASC NULLS FIRST, 200)
               +- *(1) Project [_1#3 AS a#6, _2#4 AS b#7]
                  +- *(1) SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true, false) AS _2#4]
                     +- Scan ExternalRDDScan[obj#2]
```

while after the PR produces:

```
== Parsed Logical Plan ==
'Sort ['a ASC NULLS FIRST], true
+- 'Project ['b]
   +- 'SubqueryAlias t
      +- 'Sort ['a ASC NULLS FIRST], true
         +- 'Project ['a, 'b]
            +- 'UnresolvedRelation `table1`

== Analyzed Logical Plan ==
b: string
Project [b#7]
+- Sort [a#6 ASC NULLS FIRST], true
   +- Project [b#7, a#6]
      +- SubqueryAlias t
         +- Sort [a#6 ASC NULLS FIRST], true
            +- Project [a#6, b#7]
               +- SubqueryAlias table1
                  +- Project [_1#3 AS a#6, _2#4 AS b#7]
                     +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(assertnotnull(input[0, scala.Tuple2, true]))._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(assertnotnull(input[0, scala.Tuple2, true]))._2, true, false) AS _2#4]
                        +- ExternalRDD [obj#2]

== Optimized Logical Plan ==
Project [b#7]
+- Sort [a#6 ASC NULLS FIRST], true
   +- Project [_1#3 AS a#6, _2#4 AS b#7]
      +- SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true, false) AS _2#4]
         +- ExternalRDD [obj#2]

== Physical Plan ==
*(2) Project [b#7]
+- *(2) Sort [a#6 ASC NULLS FIRST], true, 0
   +- Exchange rangepartitioning(a#6 ASC NULLS FIRST, 5)
      +- *(1) Project [_1#3 AS a#6, _2#4 AS b#7]
         +- *(1) SerializeFromObject [staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._1, true, false) AS _1#3, staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, assertnotnull(input[0, scala.Tuple2, true])._2, true, false) AS _2#4]
            +- Scan ExternalRDDScan[obj#2]
```

this means that an unnecessary sort operation is not performed after the PR.

## How was this patch tested?

added UT

mgaido91 committed:
[SPARK-23375][SQL] Eliminate unneeded Sort in Optimizer

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/752/
Test PASSed.

SparkQA commented:
**[Test build #87261 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87261/testReport)** for PR 20560 at commit [`550ff99`](https://github.com/apache/spark/commit/550ff99652de515e9ee056596350a8cbf802f938).

SparkQA commented:
**[Test build #87261 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87261/testReport)** for PR 20560 at commit [`550ff99`](https://github.com/apache/spark/commit/550ff99652de515e9ee056596350a8cbf802f938).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `abstract class KeepOrderUnaryNode extends UnaryNode `
  * `case class Subquery(child: LogicalPlan) extends KeepOrderUnaryNode `
  * `case class Project(projectList: Seq[NamedExpression], child: LogicalPlan)`
  * `case class GlobalLimit(limitExpr: Expression, child: LogicalPlan) extends KeepOrderUnaryNode `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87261/
Test PASSed.

gatorsmile commented:
Thanks! This should be added as a separate rule. It is actually to resolve the comment in https://github.com/apache/spark/pull/11480#discussion_r54840644

I did not review it carefully, but it requires more test cases, including unit tests and end-to-end tests

mgaido91 committed:
Use separate rule and add more tests

SparkQA commented:
**[Test build #87288 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87288/testReport)** for PR 20560 at commit [`81e4828`](https://github.com/apache/spark/commit/81e48286806d36e7630e961168d87cbad4f10194).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/774/
Test PASSed.

mgaido91 commented:
@gatorsmile thanks for your comment. I moved it to a separate rule and added more tests.

As per the added value of this rule, I see 3 main points:
 1. Let's imagine that a user exposes a cached sorted relation which can be queried by other users via JDBC. Other users cannot know that the table is already sorted and they may write query which cause an unnecessary sort.
 2. Many tools which produce automatic SQL code are not very smart in creating it, so they can generate queries which cause unneeded sorts.
 3. I think this is also enabling for more interesting use cases. What I am thinking about is that we may have some datasources which store sorted data and if we can express this in the logical plan, then we may avoid unneeded sorts.

What do you think?
Thanks.

SparkQA commented:
**[Test build #87288 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/87288/testReport)** for PR 20560 at commit [`81e4828`](https://github.com/apache/spark/commit/81e48286806d36e7630e961168d87cbad4f10194).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class LocalLimit(limitExpr: Expression, child: LogicalPlan) extends KeepOrderUnaryNode `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/87288/
Test PASSed.

gatorsmile commented:
@mgaido91 Yeah, we definitely should include this rule. We just need more careful review and comprehensive test cases. Thanks for your work!

mgaido91 commented:
thank you @gatorsmile for taking a look at this. Let me know if there is something I can/should improve. Thanks.

mgaido91 commented:
@gatorsmile sorry, do you have time now to take a look at this? Or may I ping you some days later if you are busy? Thanks.

gatorsmile commented:
Will review this in the next few days.

mgaido91 commented:
kindly ping @gatorsmile

gatorsmile commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -733,6 +735,17 @@ object EliminateSorts extends Rule[LogicalPlan] {
+object RemoveRedundantSorts extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Sort(orders, true, child) if child.sortedOrder.nonEmpty
+        && child.sortedOrder.zip(orders).forall { case (s1, s2) => s1.satisfies(s2) } =>
```
Why not using `SortOrder.orderingSatisfies`?

gatorsmile commented:
cc @cloud-fan @hvanhovell @wzhfy

mgaido91 committed:
rename to outputOrdering

mgaido91 commented:
retest this please

SparkQA commented:
**[Test build #88849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88849/testReport)** for PR 20560 at commit [`1c33263`](https://github.com/apache/spark/commit/1c33263c275746b25727d04e3a7ada14140b0b68).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1923/
Test PASSed.

SparkQA commented:
**[Test build #88849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88849/testReport)** for PR 20560 at commit [`1c33263`](https://github.com/apache/spark/commit/1c33263c275746b25727d04e3a7ada14140b0b68).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88849/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -733,6 +735,17 @@ object EliminateSorts extends Rule[LogicalPlan] {
+object RemoveRedundantSorts extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Sort(orders, true, child) if child.outputOrdering.nonEmpty
+        && SortOrder.orderingSatisfies(child.outputOrdering, orders) =>
```
shall we do it after planning as we already have `SparkPlan. outputOrdering`?

cloud-fan replied:
ah they are different. This is global ordering,

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -43,11 +43,12 @@ case class ReturnAnswer(child: LogicalPlan) extends UnaryNode {
 }
 
-case class Project(projectList: Seq[NamedExpression], child: LogicalPlan) extends UnaryNode {
+case class Project(projectList: Seq[NamedExpression], child: LogicalPlan)
```
Like `ProjectExec.outputOrdering`, we can propagate ordering for aliased attributes.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -867,6 +871,11 @@ case class RepartitionByExpression(
   override def shuffle: Boolean = true
+
+  override def outputOrdering: Seq[SortOrder] = partitioning match {
+    case RangePartitioning(ordering, _) => ordering
```
`RangePartitioning` doesn't guarantee ordering inside partition, we can't do this.

mgaido91 committed:
remove outputOrdering for RangePartitioning

SparkQA commented:
**[Test build #88888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88888/testReport)** for PR 20560 at commit [`60ea6fc`](https://github.com/apache/spark/commit/60ea6fcee5777d79eb3f9b6e86db1a5516995e7a).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/1957/
Test PASSed.

SparkQA commented:
**[Test build #88888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/88888/testReport)** for PR 20560 at commit [`60ea6fc`](https://github.com/apache/spark/commit/60ea6fcee5777d79eb3f9b6e86db1a5516995e7a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/88888/
Test PASSed.

wzhfy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantSortsSuite.scala:
```diff
@@ -0,0 +1,93 @@
+class RemoveRedundantSortsSuite extends PlanTest {
+  override val conf = new SQLConf().copy(CASE_SENSITIVE -> true, ORDER_BY_ORDINAL -> false)
+  val catalog = new SessionCatalog(new InMemoryCatalog, EmptyFunctionRegistry, conf)
+  val analyzer = new Analyzer(catalog, conf)
```
If we don't use ordinal number, we can remove these.

wzhfy commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantSortsSuite.scala:
```diff
@@ -0,0 +1,93 @@
+  test("remove redundant order by") {
+    val orderedPlan = testRelation.select('a, 'b).orderBy('a.asc, 'b.desc_nullsFirst)
+    val unnecessaryReordered = orderedPlan.select('a).orderBy('a.asc, 'b.desc_nullsFirst)
+    val optimized = Optimize.execute(analyzer.execute(unnecessaryReordered))
```
just use `unnecessaryReordered.analyze`?

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -733,6 +735,17 @@ object EliminateSorts extends Rule[LogicalPlan] {
 }
 
+/**
+ * Removes Sort operations on already sorted data
```
how about `Removes Sort operation if the child is already sorted`?

wzhfy commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:
```diff
@@ -522,6 +524,8 @@ case class Range(
     Statistics(sizeInBytes = LongType.defaultSize * numElements)
   }
+
+  override def outputOrdering: Seq[SortOrder] = output.map(a => SortOrder(a, Descending))
```
ordering is the same when `step` in `Range` is positive or negative?

mgaido91 replied:
Nice catch, thanks! I missed it!

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -197,6 +198,19 @@ class PlannerSuite extends SharedSQLContext {
+    val resorted = query.sort('key.desc)
+    assert(resorted.queryExecution.optimizedPlan.collect { case s: Sort => s}.isEmpty)
+    assert(resorted.select('key).collect().map(_.getInt(0)).toSeq ==
+      (1 to 100).sorted(Ordering[Int].reverse))
```
`(1 to 100).reverse`?

wzhfy commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -197,6 +198,19 @@ class PlannerSuite extends SharedSQLContext {
+      (1 to 100).sorted(Ordering[Int].reverse))
+    // with a different order, the sort is needed
+    val sortedAsc = query.sort('key)
+    assert(sortedAsc.queryExecution.optimizedPlan.collect { case s: Sort => s}.nonEmpty)
```
`.nonEmpty` -> `.size == 1`

mgaido91 committed:
fix Range ordering + add more Range UT + address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2149/
Test PASSed.

SparkQA commented:
**[Test build #89118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89118/testReport)** for PR 20560 at commit [`1c7cae6`](https://github.com/apache/spark/commit/1c7cae685314bf762b38defb9233dbef315ab0df).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89118/
Test FAILed.

mgaido91 committed:
fix ut failure

SparkQA commented:
**[Test build #89135 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89135/testReport)** for PR 20560 at commit [`e376c19`](https://github.com/apache/spark/commit/e376c193b44d5293cf9e7075b83149c93d1a9342).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2164/
Test PASSed.

SparkQA commented:
**[Test build #89135 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89135/testReport)** for PR 20560 at commit [`e376c19`](https://github.com/apache/spark/commit/e376c193b44d5293cf9e7075b83149c93d1a9342).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89135/
Test PASSed.

henryr commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -22,10 +22,11 @@ import org.apache.spark.sql.{execution, Row}
 import org.apache.spark.sql.execution.columnar.InMemoryRelation
-import org.apache.spark.sql.execution.exchange.{EnsureRequirements, ReusedExchangeExec, ReuseExchange, ShuffleExchangeExec}
+import org.apache.spark.sql.execution.exchange.{EnsureRequirements, ReusedExchangeExec, ReuseExchange,
+  ShuffleExchangeExec}
```
revert this?

mgaido91 replied:
why?

cloud-fan replied:
it's a unnecessary change. We don't have length limit for imports

henryr commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -733,6 +735,17 @@ object EliminateSorts extends Rule[LogicalPlan] {
+    case Sort(orders, true, child) if child.outputOrdering.nonEmpty
+        && SortOrder.orderingSatisfies(child.outputOrdering, orders) =>
+      child
+  }
```
You might not want to do it in this PR, but you could easily remove another simple kind of redundant sort, e.g.:

`rel.orderBy('a.desc).orderBy('a.asc)`

(and I think that `orderBy` is not stable, so any two consecutive `orderBy` operators are redundant).

mgaido91 replied:
Yes, you're right. Probably we can do this in other PR. May you open a JIRA for this? Thanks!

cloud-fan replied:
This is a good follow-up

henryr replied:
Filed SPARK-23973 for this

henryr reviewed (commented):
Just a couple suggestions, feel free to ignore.

mgaido91 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -274,3 +279,7 @@ abstract class BinaryNode extends LogicalPlan {
   override final def children: Seq[LogicalPlan] = Seq(left, right)
 }
+
+abstract class KeepOrderUnaryNode extends UnaryNode {
```
thanks for the suggestion. I'd love to hear also @cloud-fan's and @wzhfy's opinion on this in order to choose all together the best name for it. What do you think?

cloud-fan replied:
OrderPreservingUnaryNode sounds better.

It only makes sense for unary node, so I don't think mixin trait is a good idea.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:
```diff
@@ -169,4 +169,6 @@ case class InMemoryRelation(
   override protected def otherCopyArgs: Seq[AnyRef] =
     Seq(_cachedColumnBuffers, sizeInBytesStats, statsOfPlanToCache)
+
+  override def outputOrdering: Seq[SortOrder] = child.outputOrdering
```
in SparkPlan
```
/** Specifies how data is ordered in each partition. */
def outputOrdering: Seq[SortOrder] = Nil
```

So we can't do this

cloud-fan replied:
We should carry the logical ordering from the cached logical plan when building the `InMemoryRelation`

mgaido91 committed:
rename to OrderPreservingUnaryNode

mgaido91 committed:
fix InMemoryRelation

SparkQA commented:
**[Test build #89195 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89195/testReport)** for PR 20560 at commit [`a1846ab`](https://github.com/apache/spark/commit/a1846abcea98c9174b0fa06bbb2dcd5545ef6eb3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2210/
Test PASSed.

SparkQA commented:
**[Test build #89195 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89195/testReport)** for PR 20560 at commit [`a1846ab`](https://github.com/apache/spark/commit/a1846abcea98c9174b0fa06bbb2dcd5545ef6eb3).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89195/
Test FAILed.

mgaido91 committed:
Merge branch 'master' of github.com:apache/spark into SPARK-23375

mgaido91 committed:
fix build failure

SparkQA commented:
**[Test build #89197 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89197/testReport)** for PR 20560 at commit [`4e441f8`](https://github.com/apache/spark/commit/4e441f81af2ed43353224e5a82487a8889d37619).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2212/
Test PASSed.

SparkQA commented:
**[Test build #89197 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89197/testReport)** for PR 20560 at commit [`4e441f8`](https://github.com/apache/spark/commit/4e441f81af2ed43353224e5a82487a8889d37619).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89197/
Test FAILed.

mgaido91 committed:
fix UT

SparkQA commented:
**[Test build #89249 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89249/testReport)** for PR 20560 at commit [`6e95e37`](https://github.com/apache/spark/commit/6e95e37d6475cb2ede1e0a8e128d3ee96c4d3273).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89249/
Test FAILed.

mgaido91 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2255/
Test PASSed.

SparkQA commented:
**[Test build #89257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89257/testReport)** for PR 20560 at commit [`6e95e37`](https://github.com/apache/spark/commit/6e95e37d6475cb2ede1e0a8e128d3ee96c4d3273).

SparkQA commented:
**[Test build #89257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89257/testReport)** for PR 20560 at commit [`6e95e37`](https://github.com/apache/spark/commit/6e95e37d6475cb2ede1e0a8e128d3ee96c4d3273).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89257/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -733,6 +735,17 @@ object EliminateSorts extends Rule[LogicalPlan] {
+ */
+object RemoveRedundantSorts extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case Sort(orders, true, child) if child.outputOrdering.nonEmpty
```
`child.outputOrdering.nonEmpty` looks like unnecessary

mgaido91 committed:
address comment

SparkQA commented:
**[Test build #89330 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89330/testReport)** for PR 20560 at commit [`6c5f04c`](https://github.com/apache/spark/commit/6c5f04cb989736ced5d7c8695a0740e512df36c6).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/2309/
Test PASSed.

SparkQA commented:
**[Test build #89330 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89330/testReport)** for PR 20560 at commit [`6c5f04c`](https://github.com/apache/spark/commit/6c5f04cb989736ced5d7c8695a0740e512df36c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89330/
Test PASSed.

cloud-fan commented:
thanks, merging to master!

rxin commented:
Just saw this - this seems like a somewhat awkward way to do it by just matching on filter / project. Is the main thing lacking a way to do back propagation for properties? (We can only do forward propagation at the moment on properties so we can't eliminate subtree's sort based on the parent's sort).

cloud-fan commented:
@rxin  It seems you are talking about the followup PR: https://github.com/apache/spark/pull/21072

I think this is the way we do back propagation in catalyst: match a specific node, traverse down the subtree with the properties.

For forward propagation, we also need to carefully handle some nodes that would stop the propagation. In `RemoveRedundantSorts.canEliminateSort`, we are doing the same thing: only list the nodes that can retain the properties. e.g. `Limit` should stop propagating the sorting property. I think `Project`, `Filter`, `Hint` is good enough as an initial list, we can expand it later.

gatorsmile commented on sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryRelation.scala:
```diff
@@ -64,7 +64,8 @@ case class InMemoryRelation(
     val sizeInBytesStats: LongAccumulator = child.sqlContext.sparkContext.longAccumulator,
-    statsOfPlanToCache: Statistics)
+    statsOfPlanToCache: Statistics,
+    override val outputOrdering: Seq[SortOrder])
```
This should be added to `otherCopyArgs `; otherwise, we will lose it when doing the tree transformation. https://github.com/apache/spark/pull/22715 fixed it.
